### ISSUE TITLE
Separate react-rails into two gems?

### ISSUE THREAD

**Issue by justin808** [reactjs/react-rails#310]
@rmosolgo Any opinion if we could or should break react-rails up into two gems? I see the gem as 2 major parts:

1. The view helpers and corresponding parts that support server side rendering
2. Sprockets integration so that it's super easy to include react and JSX into a Rails project. This gem could include part 1) as a dependency, as well as the [react-source](https://github.com/reactjs/react-rails/blob/master/VERSIONS.md) gem.

For those that use these systems below, it would be nice not include the parts related to 2).

Webpack: https://github.com/justin808/react-webpack-rails-tutorial

browserify-rails: https://github.com/browserify-rails/browserify-rails

**Comment by vipulnsward** [comment]
@justin808 any conflicts with how its currently available?

I think both of these parts are relatively pretty small to extract into two gems.

**Comment by justin808** [comment]
@vipulnsward I'm not sure if there are any conflicts. I guess I'll fork and find out!

**Comment by rmosolgo** [comment]
What's the advantage to having two code bases vs. not using the parts of react-rails that don't help your case? Do the features in `2)` get in the way?

I also took a spin at extracting JSX a while back: https://github.com/rmosolgo/jsx-ruby I was thinking some sinatra users (etc) might want to use React.

**Comment by guzart** [comment]
Only problem I have is that I need to include sprockets in my app, and it's only because of the Engine initialization.

**Comment by justin808** [comment]
@rmosolgo Hi! After a fair bit tinkering, we've got a working version of a gem focused on using React + Rails with Webpack: https://github.com/shakacode/react_on_rails/. Any feedback would be awesome. We're going to be adding lots of utilities for making webpack integrations with Rails easier.

@guzart What do you mean "sprockets" being an issue?

**Comment by guzart** [comment]
@justin808 react-trails requires a Rails asset engine. This is because of https://github.com/reactjs/react-rails/blob/master/lib/react/rails/engine.rb#L5  Since it requires an asset engine I am forced to `require 'sprockets/railtie'` in my `application.rb` which means that files are being served at `http://localhost:3000/assets/application.js`.

My goal here is to get rid of sprockets completely and use webpack as my asset manager.

**Comment by justin808** [comment]
@guzart Why do you want to do that? One of my team members at https://github.com/shakacode/ is @alexfedoseev who wrote up: https://medium.com/@alexfedoseev/isomorphic-react-with-rails-part-i-440754e82a59.

There's many advantages to sticking with sprockets, mainly around simplicity, such as:

1. Using one port for the app so not needing to setup CORS.
2. Let Rails handle image/font fingerprinting for use with sass-rails, and bootstrap-sass.

**Comment by guzart** [comment]
@justin808 That's a nice post, but adding an express adds complexity to the app. Which for some might be a good tradeoff in exchange for performance on server side rendering but it's not for me.

The only two reasons I want to use webpack instead of sprockets is because I want source maps and react hot-reload for my DX. Of course you could have a hybrid, which I do right now but it's complex since it requires you to know about both technologies and their boundaries.

I agree that using sprockets is simple and it works out of the box. But that's what am trying to build right now, a gem that'll let you have react hot loader and server side rendering out of the box and with simple configuration.

    1. Setting up CORS is a one line configuration on the Webpack dev server
    2. Webpack was built for asset management, and there are plugins for fingerprinting assets and generating an asset manifest. Once you fully move to webpack there are npm libraries for bootstrap and sass.

**Comment by justin808** [comment]
@guzart wrote:
> but adding an express adds complexity to the app. Which for some might be a good tradeoff in exchange for performance on server side rendering but it's not for me.

The "hot reload server" setup on my system on a different port is totally, 100% optional, if you want to see live reloading. On Alex's system, it's a lot of extra work. FWIW, Alex and I are working on a production system and we're using react_on_rails.

A year ago, I wrote the https://www.npmjs.com/package/bootstrap-sass-loader/ and I experimented with the extract-text-plugin to write out the sass, but it was more work to make it work with asset fingerprinting, while the bootstrap-sass gem with sass-rails worked perfectly. So we settled on a solution that uses a few shared files and a few sym links.

In any case, I'm looking forward to seeing what you come up with and any feedback on our https://github.com/shakacode/react_on_rails gem!
